core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java

```diff
@@ -0,0 +1,659 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.AlluxioURI;
+import alluxio.Configuration;
+import alluxio.Constants;
+import alluxio.PropertyKey;
+import alluxio.underfs.options.CreateOptions;
+import alluxio.underfs.options.DeleteOptions;
+import alluxio.underfs.options.FileLocationOptions;
+import alluxio.underfs.options.ListOptions;
+import alluxio.underfs.options.MkdirsOptions;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.util.CommonUtils;
+import alluxio.util.io.PathUtils;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A object based abstract {@link UnderFileSystem}.
+ */
+@ThreadSafe
+public abstract class ObjectUnderFileSystem extends BaseUnderFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  /** Maximum length for a single listing query. */
+  private static final int LISTING_LENGTH_MAX = 1000;
+
+  /** Length of each list request. */
+  protected static final int LISTING_LENGTH =
+      Configuration.getInt(PropertyKey.UNDERFS_LISTING_LENGTH) > LISTING_LENGTH_MAX
+          ? LISTING_LENGTH_MAX : Configuration.getInt(PropertyKey.UNDERFS_LISTING_LENGTH);
+
+  /** Value used to indicate nested structure. */
+  protected static final char PATH_SEPARATOR_CHAR = '/';
+
+  /** Value used to indicate nested structure. */
+  protected static final String PATH_SEPARATOR = String.valueOf(PATH_SEPARATOR_CHAR);
+
+  /**
+   * Constructs an {@link ObjectUnderFileSystem}.
+   *
+   * @param uri the {@link AlluxioURI} used to create this ufs
+   */
+  protected ObjectUnderFileSystem(AlluxioURI uri) {
+    super(uri);
+  }
+
+  /**
+   * Information about a single object in object UFS.
+   */
+  protected class ObjectStatus {
+    long mFileSize;
+    long mLastModifiedTimeMs;
+
+    public ObjectStatus(long fileSize, long lastModifiedTimeMs) {
+      mFileSize = fileSize;
+      mLastModifiedTimeMs = lastModifiedTimeMs;
+    }
+
+    /**
+     * Gets the file size in bytes.
+     *
+     * @return the file size in bytes
+     */
+    public long getContentLength() {
```

Comment:
Well it is consistent with the terminology used by object stores as they do not refer to 'objects' as 'files'. I renamed the variable to be consistent. I can change it if you feel strongly about it.